### PR TITLE
Feature/missing samples

### PR DESCRIPTION
**Pull request by Nuna7**
Issue #1736

1. Add functionality to simulate missing values for BayesianNetwork.
2. Add test cases for the new feature
3. Add relative tolerance for test case, test_connected_predict_probability, in test_BayesianModel.py

### CODE REVIEW

**Comment by ankurankan** [comment]
@Nuna7 Thanks for the PR. This issue has turned out to be much more complicated than I had originally thought when creating the issue. I have just added a reference in the issue to the paper discussing different types of missingness using DAGs. I think your current implementation covers the MCAR (missing completely at random) case, so should still be useful. For the other cases, I think we will need to first figure out how to represent those.

File: pgmpy/models/BayesianNetwork.py [hunk 1]
```diff
@@ -1227,6 +1230,19 @@ def simulate(
...
+            If True, include missing values in the samples.
+
+        missing_prob: float
+            The probability that there is missing values in the samples.
+
+        missing_columns: list
+            The list of columns where there will be missing values in the samples.
+            If None, then all columns could contain the missing values.
+
+        Simulation with missing values:
```
Comment (ankurankan) [inline pgmpy/models/BayesianNetwork.py]: Could you please move the example to the `Examples` section in the docstring.

File: pgmpy/models/BayesianNetwork.py [hunk 2]
```diff
@@ -1346,7 +1362,27 @@ def simulate(
...
 
-        # Step 5: Postprocess and return
+        # Step 5: If include_missing; include missing values in samples.
+        if include_missing:
+            if missing_prob <= 0:
+                raise ValueError("Missingness probability should be greater than 0")
+            if missing_prob >= 1:
+                raise ValueError("Missingness probability should be less than 1")
+
+            mask = np.random.rand(*samples.shape)
```
Comment (ankurankan) [inline pgmpy/models/BayesianNetwork.py]: Could you use numpy's generator interface here (https://numpy.org/doc/stable/reference/random/generator.html) ? And also set a seed value if passed in the argument, to make this reproducible over multiple runs.

File: pgmpy/models/BayesianNetwork.py [hunk 3]
```diff
@@ -1227,6 +1230,19 @@ def simulate(
         show_progress: bool
             If True, shows a progress bar when generating samples.
 
+        include_missing: bool
+            If True, include missing values in the samples.
+
+        missing_prob: float
+            The probability that there is missing values in the samples.
+
+        missing_columns: list
```
Comment (ankurankan) [inline pgmpy/models/BayesianNetwork.py]: Please specify the default values here.

File: pgmpy/tests/test_sampling/test_Sampling.py [hunk 4]
```diff
@@ -374,6 +374,34 @@ def test_rejection_sample_basic(self):
         self.assertTrue(set(sample.G).issubset({"g0", "g1"}))
         self.assertTrue(set(sample.L).issubset({"l0", "l1"}))
 
+    def test_missing_scheme_sample(self):
```
Comment (ankurankan) [inline pgmpy/tests/test_sampling/test_Sampling.py]: Could you move this test to the `test_BayesianNetworks.py` file as the feature is implemented in BayesianNetwork.py?

**Review by ankurankan — CHANGES_REQUESTED** [review]
The PR looks good to me. I have added a few minor comments.

**Comment by ankurankan** [comment]
@Nuna7 BTW you can install a pre-commit hook to automatically fix all code formatting when committing. From the pgmpy dir, you can run:
```
pip install pre-commit
pre-commit install
```

**Comment by Nuna7** [comment]
> @Nuna7 BTW you can install a pre-commit hook to automatically fix all code formatting when committing. From the pgmpy dir, you can run:
> 
> ```
> pip install pre-commit
> pre-commit install
> ```

**Comment by ankurankan** [comment]
@Nuna7 Any reason for closing this? I thought this was mostly ready to be merged.

**Comment by Nuna7** [comment]
@ankurankan Initially, I thought you closed it (being a beginner in pull request) and just now I realized that I may mistakenly closed it.

**Review by ankurankan — APPROVED** [review]

**Comment by ankurankan** [comment]
@Nuna7 Thanks for the PR :)

**Pull request merged**